src/Conversion/ONNXToTOSA/ONNXToTOSACommon.cpp

```diff
@@ -30,6 +31,274 @@ using namespace mlir;
+      inputIdxIndices.push_back(i);
+    } else {
+      inputRightChannels.push_back(inputShape[i]);
+      inputIdxRightChannels.push_back(i);
+    }
+  }
+
+  // Calculate N, K, W, C
+  int64_t N = multiplyDims(inputShape.take_front(batchDims));
+  int64_t W =
+      multiplyDims(indicesShape.slice(batchDims, indicesRank - batchDims));
+  int64_t K = inputShape[axis];
+
+  int64_t C = multiplyDims(inputShape.slice(batchDims, axis - batchDims));
+  C = multiplyDims(inputShape.slice(axis + 1, inputRank - axis - 1), C);
+
+  /////////////////////////////////////////////
+  // Build up the input transpose operator
+  SmallVector<int32_t> inputTransposePerm;
+
+  // Batch
+  for (size_t i = 0; i < inputBatch.size(); i++) {
+    inputTransposePerm.push_back(inputIdxBatch[i]);
+  }
+
+  // Indices
+  for (size_t i = 0; i < inputIndices.size(); i++) {
+    inputTransposePerm.push_back(inputIdxIndices[i]);
+  }
+
+  // LeftChannels
+  for (size_t i = 0; i < inputLeftChannels.size(); i++) {
+    inputTransposePerm.push_back(inputIdxLeftChannels[i]);
+  }
+
+  // RightChannels
+  for (size_t i = 0; i < inputRightChannels.size(); i++) {
+    inputTransposePerm.push_back(inputIdxRightChannels[i]);
+  }
+
+  /////////////////////////////////////////////
+  // Build up the result reshape, in prepration for transpose
+  // [N, W, C] -> [ Batch, Indices, LeftChannels, RightChannels ]
+  SmallVector<int64_t> resultReshapeShape;
+
+  // Indices
+  for (size_t i = 0; i < indicesType.getShape().size(); i++) {
+    resultReshapeShape.push_back(indicesShape[i]);
+  }
+
+  // Left channels
+  for (size_t i = 0; i < inputLeftChannels.size(); i++) {
+    resultReshapeShape.push_back(inputLeftChannels[i]);
+  }
+
+  // Right channels.  But remove the axis dimension.
+  for (size_t i = 0; i < inputRightChannels.size(); i++) {
+    resultReshapeShape.push_back(inputRightChannels[i]);
+  }
+
+  /////////////////////////////////////////////
+  // Build up the result transpose operator.
+  SmallVector<int32_t> resultTransposePerm;
+
+  // Batch dimensions
+  for (int i = 0; i < batchDims; i++) {
+    resultTransposePerm.push_back(i);
+  }
+
+  // LeftChannels
+  for (int i = 0; i < (int)inputLeftChannels.size(); i++) {
+    resultTransposePerm.push_back(i + indicesType.getShape().size());
+  }
+
+  // Indices (remainder of dimensions after batch).
+  for (int i = batchDims; i < (int)(indicesType.getShape().size()); i++) {
+    resultTransposePerm.push_back(i);
+  }
+
+  // RightChannels, coming from after both the Indices and LeftChannels.
+  for (int i = 0; i < (int)inputRightChannels.size(); i++) {
+    resultTransposePerm.push_back(
+        i + indicesType.getShape().size() + inputLeftChannels.size());
+  }
+
+  SmallVector<int64_t> tosaValuesShape = {N, K, C};
+  SmallVector<int64_t> tosaIndicesShape = {N, W};
+
+  // Begin of rewrite.
+
+  auto inputTransposeOp = tosaBuilder.transpose(inputValue, inputTransposePerm);
+
+  if (countDynamicDims(tosaValuesShape) > 1) {
```

Comment:
This is your first mention of supporting dynamic shapes. Where does the restriction of only one from `{N, K, C}` and only one from `{N, W}` come from, and can it really be any from those sets?

Reply:
This comes form the[ reshape op of TOSA](https://www.mlplatform.org/tosa/tosa_spec.html#_reshape). One Dimension can be dynamic because the shape can be calculated.